subprojects/core/src/main/java/org/gradle/api/internal/tasks/DefaultTaskContainer.java

```diff
@@ -274,12 +274,15 @@ public Task create(String name, Closure configureClosure) {
         return Cast.uncheckedCast(createLater(name, DefaultTask.class, configurationAction));
     }
 
-    public <T extends Task> TaskProvider<T> createLater(final String name, final Class<T> type, @Nullable  Action<? super T> configurationAction) {
+    public <T extends Task> TaskProvider<T> createLater(final String name, final Class<T> type, @Nullable Action<? super T> configurationAction) {
         if (hasWithName(name)) {
             duplicateTask(name);
         }
-        DefaultTaskProvider<T> provider = new TaskCreatingProvider<T>(type, name, configurationAction);
+        DefaultTaskProvider<T> provider = new TaskCreatingProvider<T>(type, name);
         addLater(provider);
+        if (configurationAction != null) {
+            provider.configure(configurationAction);
```

Comment:
Deferring the configuration like this has made another problem worse.  That's why some integration tests are failing.  Imagine:

```
tasks.createLater("myTask", MyTask) { 
   someProperty = "foo"
}
tasks.create(name: "myTask", type: DefaultTask, overwrite: true) {
   doLast { println "I am no longer a MyTask" }
}
```
I think previously, the configuration would just be lost when the new task overwrote it.  Now we try to execute the configuration block against the new task, which happens to be a different type.

I think this is similar to what you could mindlessly do before:
```
tasks.create("myTask", MyTask)
tasks.create(name: "myTask", type: DefaultTask, overwrite: true)
tasks.matching { it.name == "myTask" }.all { 
   someProperty = "foo" // BOOM
}
```

Reply:
And it looks like the placeholders have a similar problem.  We create them lazily with the proper type, but then we create them with a different type eagerly and the pre-existing configuration action executes against the wrong type.